1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,154 @@
+Imported modules and function definitions are not affected.
+A script that defines a constant function would fail if the function is already defined.  
+The initial assumption will be that runspace/thread creation time is insignificant compared to the time needed to execute the script block, either because of high compute needs or because of long wait times for results.
+If this assumption is not true then the user should consider batching the work load to each foreach-object iteration, or simply use the sequential/non-parallel form of the cmdlet.
+
+The 'TimeoutSecs' parameter will attempt to halt all script block executions after the timeout time has passed, however it may not be immediately successful if the running script is calling a native command or API, in which case it needs for the call to return before it can halt the running script.
+
+### Variable passing
+
+ForEach-Object -Parallel will support the PowerShell `$_` current piped item variable within each script block.
+It will also support the `$using:` directive for passing variables from script scope into the parallel executed script block scope.  
+If the passed in variable is a value type, a copy of the value is passed to the script block.  
+If the passed in variable is a reference type, the reference is passed and each running script block can modify it.
+Since the script blocks are running in different threads, modifying a reference type that is not thread safe will result in undefined behavior.
+
+### Supported scenarios
+
+```powershell
+# Ensure needed module is installed on local system
+if (! (Get-Module -Name MyLogsModule -ListAvailable)) {
+    Install-Module -Name MyLogsModule -Force
+}
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$logs = $computerNames | ForEach-Object -Parallel -ThrottleLimit 10 -TimeoutSecs 1800 -ScriptBlock {
+    Get-Logs -ComputerName $_
+}
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$job = ForEach-Object -Parallel -ThrottleLimit 10 -InputObject $computerNames -TimeoutSecs 1800 -AsJob -ScriptBlock {
+    Get-Logs -ComputerName $_
+}
+$logs = $job | Wait-Job | Receive-Job
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$logNames = 'System','SQL'
+$logs = ForEach-Object -Parallel -InputObject $computerNames -ScriptBlock {
+    Get-Logs -ComputerName $_ -LogNames $using:logNames
+}
+```
+
+```powershell
+$computerNames = 'computer1','computer2','computer3','computer4','computer5'
+$logNames = 'System','SQL','AD','IIS'
+$logResults = ForEach-Object -Parallel -InputObject $computerNames -ScriptBlock {
+    Get-Logs -ComputerName $_ -LogNames $using:logNames
+} | ForEach-Object -Parallel -ScriptBlock {
+    Process-Log $_
+}
+```
+
+### Unsupported scenarios

```

Comment:
Looking on the unsupported scenarios I see an inconsistence - we can send variables by `$using` but can return result only by `return`.  Maybe discard `$using` and only use `-ArgumentList` and `param`? It seems it will be more clear for users (one entry point - one exit point) and it will be one step closer to being thread-safe. It also resolve problems like "Value types not passed by reference"/"Passed in reference variables should not be assigned to".

Reply:
Start-Parallel (https://www.powershellgallery.com/packages/start-parallel/1.3.0.0) goes down the `-argumentlist ` and `param()`  route.  

`$using` variables are read-only copies.. When you run a script block on on another computer I don't think there is any expectation that it sets variables on the computer that invokes the command.  And in DSC a `Using` variable is written into the MOF file,  -  when DSC applies the MOF file it wouldn't make sense to try to go back to the runspace which compiled it and change a variable - the application phase will often happen when that runspace has ceased to exist.   
Changing form the the _read-only copy_ paradigm means 
(a) The runspace using the variable must read the current value "live" from the calling runspace
(b) The child runspace must have a mechanism to write back, and the parent must track the variables the each runspace can read and write and authenticate that whatever is asking to read or write is really that runspace
(c) because there may be multiple runspaces writing back to the same variable there must be some locking mechanism. One runspace locking a variable while it performs multiple updates may block others.  

One of the things I keep saying is if the desire is to make a parallel version of `foreach-object`  then: 
1. Differences from existing sequential foreach should be avoided where possible (hence keeping `begin `and `end `blocks; returning _jobs_ very much optional).
2. Many existing script blocks will fail if run  _unmodified_ in parallel runspaces.  Therefore one difference is to **reserve `-process`   for _sequential_ use.** i.e. changing ` Foreach-object {$variable} `  to  `Foreach-object {stuff $variable} -parallel` should not be syntactically valid. Requiring  `Foreach-object -parallel {...} ` tells us when reading or writing the code that the block runs in its own runspace with read only access to local variables, only if `$Using` is specified.
3. Switching to` Param() `and `-argumentList`  means a greater change to the scriptblock than supporting `$using `but the parameter changes when the block runs in a different runspace, try this:
```
$foo = get-item .
invoke-command -ScriptBlock {param ($foo) $foo | gm }  -ArgumentList $foo  -ComputerName localhost -EnableNetworkAccess 

invoke-command -ScriptBlock {param ($foo) $foo | gm }  -ArgumentList $foo 
```
The first invoke has a deserialized object without the methods associated with a directory. The second has a "normal" directory object. 

4. Things are complicated by `foreach-object` already supporting `-argumentlist` when used with `-memberName` e.g. 
` ForEach-Object -MemberName Split -ArgumentList "."  `
But it does not work with a `-process` block, whether it should is another matter.

Reply:
The parser does add some restrictions to $using variables, mainly because they were used for remoting.  But I think we want those restrictions given undefined behavior of assignable $using variables.

A new parameter set was added to foreach-object specifically to indicate that the -Parallel operation is not the same as the sequential operation using the traditional parameter set.  I thought about creating a new cmdlet, but I feel a new parameter set is sufficient to differentiate.